api_server.go

```diff
@@ -0,0 +1,196 @@
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"html/template"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type ApiServer struct {
+	ObjectStore
+	mux *http.ServeMux
+}
+
+type APIHandlerFunc func(r *http.Request) ([]byte, int)
+
+func (x APIHandlerFunc) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	body, code := logRequest(x, r)
+	w.WriteHeader(code)
+	w.Write(body)
+}
+
+func logRequest(handlerFunc APIHandlerFunc, r *http.Request) ([]byte, int) {
+	start := time.Now()
+	body, code := handlerFunc(r)
+
+	clientIP := strings.Split(r.Header.Get("X-Forwarded-For"), ",")[0]
+	if clientIP == "" {
+		clientIP, _, _ = net.SplitHostPort(r.RemoteAddr)
+	}
+	fields := logrus.Fields{
+		"client_ip":       clientIP,
+		"request_path":    r.URL.EscapedPath(),
+		"user_agent":      r.UserAgent(),
+		"request_method":  r.Method,
+		"request_size":    r.ContentLength,
+		"request_seconds": time.Since(start).Seconds(),
+		"status_code":     code,
+	}
+	if code < 500 {
+		logger.WithFields(fields).Info("request completed")
+	} else {
+		logger.WithFields(fields).Error("request failed")
+	}
+	return body, code
+}
+
+func (x *ApiServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if x.mux == nil {
+		x.mux = http.NewServeMux()
+		x.mux.Handle("/music_pdfs", APIHandlerFunc(x.MusicPDFsIndex))
+		x.mux.Handle("/music_pdfs/upload", APIHandlerFunc(x.MusicPDFsUpload))
+	}
+	x.mux.ServeHTTP(w, r)
+}
+
+func (x *ApiServer) MusicPDFsIndex(r *http.Request) ([]byte, int) {
+	// only accept get
+	if r.Method != http.MethodGet {
+		return nil, http.StatusMethodNotAllowed
+	}
+
+	objects := x.ListObjects(MusicPdfsBucketName)
+	pdfs := make([]MusicPDFMeta, 0, len(objects))
+	for i := range objects {
+		pdfs = append(pdfs, NewMusicPDFMetaFromTags(objects[i].Tags))
+	}
+
+	var buf bytes.Buffer
+	switch true {
+	case acceptsType(r, "text/html"):
+		musicPDFsTemplate, err := template.ParseFiles("public/music_pdfs.gohtml")
+		if err != nil {
+			logger.WithError(err).Error("template.ParseFiles() failed")
+			return nil, http.StatusInternalServerError
+		}
+		if err := musicPDFsTemplate.Execute(&buf, &pdfs); err != nil {
+			logger.WithError(err).Error("template.Execute() failed")
+			return nil, http.StatusInternalServerError
+		}
+	default:
+		if err := json.NewEncoder(&buf).Encode(&objects); err != nil {
+			logger.WithError(err).Error("json.Encode() failed")
+			return nil, http.StatusInternalServerError
+		}
+	}
+	return buf.Bytes(), http.StatusOK
+}
+
+func acceptsType(r *http.Request, mimeType string) bool {
+	for _, t := range strings.Split(r.Header.Get("Accept"), ",") {
+		if t == mimeType {
+			return true
+		}
+	}
+	return false
+}
+
+func (x *ApiServer) MusicPDFsUpload(r *http.Request) ([]byte, int) {
+	// only accept post
+	if r.Method != http.MethodPost {
+		return nil, http.StatusMethodNotAllowed
+	}
+
+	// only allow <1MB
+	if r.ContentLength > int64(1e6) {
```

Comment:
This should probably be configurable eventually. Though it's fine for a prototype. 

Is there a particular reason to choose 1 MB as the max file size? Is this a restriction of what the file store will have?

Reply:
Object storage can handle more.

Honestly, I'm not too against 1MB, because realistically, exported sheets should be kilobytes scale. We can factor this out to envvar conf later, as per 12 factor.

Reply:
Oh I just picked a number there.